cf-custom-resources/lib/custom-domain.js

```diff
@@ -0,0 +1,355 @@
+ * @param {string} lbDNS DNS of the load balancer
+ * @param {string} lbHostedZone Hosted Zone of the load balancer
+ * @param {string} rootDnsRole the IAM role ARN that can manage domainName
+ * @param {string} aliasTypes the alias type
+ */
+const writeCustomDomainRecord = async function (
+  aliases,
+  lbDNS,
+  lbHostedZone,
+  rootDnsRole,
+  aliasTypes,
+  action
+) {
+  const envRoute53 = new aws.Route53();
+  const appRoute53 = new aws.Route53({
+    credentials: new aws.ChainableTemporaryCredentials({
+      params: { RoleArn: rootDnsRole },
+      masterCredentials: new aws.EnvironmentCredentials("AWS"),
+    }),
+  });
+  if (waiter) {
+    // Used by the test suite, since waiters aren't mockable yet
+    envRoute53.waitFor = appRoute53.waitFor = waiter;
+  }
+  for (const alias of aliases) {
+    const aliasType = await getAliasType(aliasTypes, alias);
+    switch (aliasType) {
+      case aliasTypes.EnvDomainZone:
+        await writeARecord(
+          envRoute53,
+          alias,
+          lbDNS,
+          lbHostedZone,
+          aliasType.domain,
+          action
+        );
+        break;
+      case aliasTypes.AppDomainZone:
+        await writeARecord(
+          appRoute53,
+          alias,
+          lbDNS,
+          lbHostedZone,
+          aliasType.domain,
+          action
+        );
+        break;
+      case aliasTypes.RootDomainZone:
+        await writeARecord(
+          appRoute53,
+          alias,
+          lbDNS,
+          lbHostedZone,
+          aliasType.domain,
+          action
+        );
+        break;
+      // We'll skip if it is the other alias type since it will be in another account's route53.
+      default:
+    }
+  }
+};
+
+const writeARecord = async function (
+  route53,
+  alias,
+  lbDNS,
+  lbHostedZone,
+  domain,
+  action
+) {
+  let hostedZoneId = hostedZoneCache.get(domain);
+  if (!hostedZoneId) {
+    const hostedZones = await route53
+      .listHostedZonesByName({
+        DNSName: domain,
+        MaxItems: "1",
+      })
+      .promise();
+
+    if (!hostedZones.HostedZones || hostedZones.HostedZones.length == 0) {
+      throw new Error(`Couldn't find any Hosted Zone with DNS name ${domain}.`);
+    }
+    hostedZoneId = hostedZones.HostedZones[0].Id.split("/").pop();
+    hostedZoneCache.set(domain, hostedZoneId);
+  }
+  console.log(`${action} A record into Hosted Zone ${hostedZoneId}`);
+  const changeBatch = await updateRecords(
+    route53,
+    hostedZoneId,
+    action,
+    alias,
+    lbDNS,
+    lbHostedZone
+  );
+  await waitForRecordChange(route53, changeBatch.ChangeInfo.Id);
+};
+
+/**
+ * Custom domain handler, invoked by Lambda.
+ */
+exports.handler = async function (event, context) {
+  var responseData = {};
+  var physicalResourceId;
+  const props = event.ResourceProperties;
+  const [app, env, domain] = [props.AppName, props.EnvName, props.DomainName];
+  var aliasTypes = {
+    EnvDomainZone: {
+      regex: `.*${env}.${app}.${domain}`,
```

Comment:
Is it worthwhile to do `(.+\.)?{$env}.${app}.${domain}`?

Reply:
Good point! I'll do `^([^\.]+\.)?${env}.${app}.${domain}`

Reply:
The reason is because we want the `{$env}.${app}.${domain}` to be only able to handle this subdomain. We don't want it to handle something like `foo.bar.{$env}.${app}.${domain}` since that would require another subdomain delegation.